Security/Guard/JWTTokenAuthenticator.php

```diff
@@ -0,0 +1,192 @@
+     * @param TokenExtractorInterface  $tokenExtractor
+     */
+    public function __construct(
+        JWTTokenManagerInterface $jwtManager,
+        EventDispatcherInterface $dispatcher,
+        TokenExtractorInterface $tokenExtractor
+    ) {
+        $this->jwtManager     = $jwtManager;
+        $this->dispatcher     = $dispatcher;
+        $this->tokenExtractor = $tokenExtractor;
+    }
+
+    /**
+     * Returns a decoded JWT token extracted from a request.
+     *
+     * {@inheritdoc}
+     *
+     * @return PreAuthenticationJWTUserToken
+     *
+     * @throws JWTAuthenticationException If the request token cannot be decoded
+     */
+    public function getCredentials(Request $request)
+    {
+        if (false === ($jsonWebToken = $this->tokenExtractor->extract($request))) {
+            return;
+        }
+
+        $preAuthToken = new PreAuthenticationJWTUserToken($jsonWebToken);
+
+        try {
+            if (!$payload = $this->jwtManager->decode($preAuthToken)) {
+                throw JWTAuthenticationException::invalidToken();
+            }
+
+            $preAuthToken->setPayload($payload);
+        } catch (JWTDecodeFailureException $e) {
+            throw JWTAuthenticationException::invalidToken($e);
+        }
+
+        return $preAuthToken;
+    }
+
+    /**
+     * Returns an user object loaded from a JWT token.
+     *
+     * {@inheritdoc}
+     *
+     * @param PreAuthenticationJWTUserToken Implementation of the (Security) TokenInterface
+     *
+     * @throws JWTAuthenticationException If no user can be loaded from the decoded token
+     */
+    public function getUser($preAuthToken, UserProviderInterface $userProvider)
+    {
+        if (!$preAuthToken instanceof PreAuthenticationJWTUserToken) {
+            throw new \InvalidArgumentException(
+                sprintf('The first argument of the "%s()" method must be an instance of "%s".', __METHOD__, PreAuthenticationJWTUserToken::class)
+            );
+        }
+
+        $payload       = $preAuthToken->getPayload();
+        $identityField = $this->jwtManager->getUserIdentityField();
+
+        if (!isset($payload[$identityField])) {
+            throw JWTAuthenticationException::invalidPayload(
+                sprintf('Unable to find a key corresponding to the configured user_identity_field ("%s") in the token payload.', $identityField)
+            );
+        }
+
+        $identity = $payload[$identityField];
+
+        try {
+            $user = $userProvider->loadUserByUsername($identity);
+        } catch (UsernameNotFoundException $e) {
+            throw JWTAuthenticationException::invalidUser($identity, $identityField);
+        }
+
+        $authToken = new JWTUserToken($user->getRoles());
+        $authToken->setUser($user);
+        $authToken->setRawToken($preAuthToken->getCredentials());
+
+        $this->dispatcher->dispatch(Events::JWT_AUTHENTICATED, new JWTAuthenticatedEvent($payload, $authToken));
```

Comment:
I like the idea of dispatching events in the authenticator as an extension point :)